metagamer/agents/dqagent.py

```diff
@@ -0,0 +1,302 @@
+    """
+    memory to save the state, action, reward sequence from the current episode.
+    """
+
+    def __init__(self, len):
+        self.rewards = collections.deque(maxlen=len)
+        self.state = collections.deque(maxlen=len)
+        self.action = collections.deque(maxlen=len)
+        self.is_done = collections.deque(maxlen=len)
+
+    def update(self, state, action, reward):
+        # if the episode is finished we do not save to new state. Otherwise we have more states per episode than rewards
+        # and actions whcih leads to a mismatch when we sample from memory.
+        self.state.append(state)
+        self.action.append(action)
+        self.rewards.append(reward)
+        self.is_done.append(False)
+
+    def sample(self, batch_size):
+        """
+        sample "batch_size" many (state, action, reward, next state, is_done) datapoints.
+        """
+        n = len(self.is_done)
+        idx = random.sample(range(0, n - 1), batch_size)
+
+        return (
+            torch.Tensor(self.state)[idx].to(device),
+            torch.LongTensor(self.action)[idx].to(device),
+            torch.Tensor(self.state)[1 + np.array(idx)].to(device),
+            torch.Tensor(self.rewards)[idx].to(device),
+            torch.Tensor(self.is_done)[idx].to(device),
+        )
+
+    def reset(self):
+        self.rewards.clear()
+        self.state.clear()
+        self.action.clear()
+        self.is_done.clear()
+
+
+def evaluate(Qmodel, env, repeats):
+    """
+    Runs a greedy policy with respect to the current Q-Network for "repeats" many episodes. Returns the average
+    episode reward.
+    """
+    Qmodel.eval()
+    perform = 0
+    for _ in range(repeats):
+        state = env.reset()
+        done = False
+        while not done:
+            state = torch.Tensor(state).to(device)
+            with torch.no_grad():
+                values = Qmodel(state)
+            action = np.argmax(values.cpu().numpy())
+            state, reward, done, _ = env.step(action)
+            perform += reward
+    Qmodel.train()
+    return perform / repeats
+
+
+def update_parameters(current_model, target_model):
+    target_model.load_state_dict(current_model.state_dict())
+
+
+def train_network(batch_size, current, target, optim, memory, gamma):
+
+    states, actions, next_states, rewards, is_done = memory.sample(batch_size)
+
+    q_values = current(states)
+
+    next_q_values = current(next_states)
+    next_q_state_values = target(next_states)
+
+    q_value = q_values.gather(1, actions.unsqueeze(1)).squeeze(1)
+    next_q_value = next_q_state_values.gather(
+        1, torch.max(next_q_values, 1)[1].unsqueeze(1)
+    ).squeeze(1)
+    expected_q_value = rewards + gamma * next_q_value * (1 - is_done)
+
+    loss = (q_value - expected_q_value.detach()).pow(2).mean()
+
+    optim.zero_grad()
+    loss.backward()
+    optim.step()
+
+
+class DQTicTacTable(Agent):
+    def wrapper(self, env):
+        return QtableWrapper(env)
```

Comment:
This is your main problem. The idea behind a wrapper was to convert observations into the right space. QtableWrapper is converting observations into strings when you want it to be converted into a tensor (perhaps one  hot?).

You need a DQWrapper with the correct observation and action space defined.